Making sure all tensors in `torch.cat` sequence have the same dtype.

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35150 Making sure all tensors in `torch.cat` sequence have the same dtype.**

Fixes #35014

Differential Revision: [D20578589](https://our.internmc.facebook.com/intern/diff/D20578589)

Commit:
Making sure all tensors in `torch.cat` sequence have the same dtype.

Fixes #35014

gchanan commented on test/test_torch.py:
```diff
@@ -15003,6 +15003,33 @@ def test_cat_big(self, device):
         self.assertEqual(result.size(0), SIZE1 + SIZE2)
 
+    @onlyCPU
+    def test_cat_bad_dtypes(self, device):
```
why is this onlyCPU?

z-a-f replied:
Oh I was testing it locally w/o GPU -- forgot to change back

Commit:
Update on "Making sure all tensors in `torch.cat` sequence have the same dtype."


Fixes #35014

Differential Revision: [D20578589](https://our.internmc.facebook.com/intern/diff/D20578589)

Commit:
Update on "Making sure all tensors in `torch.cat` sequence have the same dtype."


Fixes #35014

Differential Revision: [D20578589](https://our.internmc.facebook.com/intern/diff/D20578589)

z-a-f commented:
Local run results:

```console
$> python test/test_torch.py -- TestTorchDeviceTypeCUDA.test_cat_bad_dtypes_cuda TestTorchDeviceTypeCPU.test_cat_bad_dtypes_cpu
..
----------------------------------------------------------------------
Ran 2 tests in 0.735s

OK
```

VitalyFedyunin commented:
@gchanan we need it in 1.5

Commit:
Update on "Making sure all tensors in `torch.cat` sequence have the same dtype."


Fixes #35014

Differential Revision: [D20578589](https://our.internmc.facebook.com/intern/diff/D20578589)
